pkg/apis/kubeone/types.go

```diff
@@ -301,6 +303,15 @@ type CanalSpec struct {
 	MTU int `json:"mtu,omitempty"`
 }
 
+// CiliumSpec defines the Cilium CNI plugin
+type CiliumSpec struct {
+	// Disable kube-proxy to use Cilium eBPF replacement
+	// default value is false
+	DisableKubeProxy bool `json:"disableKubeProxy,omitempty"`
```

Comment:
This sort of conflicts with the KubeProxy struct that we have in the ClusterNetworkConfig struct. For example, what if someone configures kube-proxy in IPVS mode via ClusterNetworkConfig, but disable it here? We should have a validation rule to reject KubeOneCluster config if this option is enabled, but the kube-proxy configuration is provided.

I'm also wondering should we put this option in the KubeProxyConfig struct instead. It might make sense to have all kube-proxy options grouped in one place. We could eventually have a validation rule to reject disabling kube-proxy if CNI is not Cilium.

Reply:
I agree on the Validation.

We could to the following:
1. It is possible to skip installation of kube-proxy for any reason.
2. Cilium needs to be configured in order to run in "without kube-proxy mode"

We need to add proper validation for this as well and it increases complexity a bit. In general i am fine to do this, however not sure if this is needed.

Reply:
I'm okay with going with the first approach — allow disabling kube-proxy for any reason regardless of the CNI. Out of curiosity, what would happen if someone disables kube-proxy and deploy Cilium, but then afterward enable kube-proxy? Would that break the cluster?

Reply:
Works fine if cilium gets configured to use kube-proxy again. But if you go the other way it will break the cluster. When kube-proxy got installed once it creates iptable rules that require cleanup on each node. For more information: https://docs.cilium.io/en/v1.10/gettingstarted/kubeproxy-free/